python/src/aiconfig/editor/client/src/components/prompt/ModelSelector.tsx

```diff
@@ -0,0 +1,82 @@
+import { Autocomplete, AutocompleteItem, Button } from "@mantine/core";
+import { memo, useState } from "react";
+import { getPromptModelName } from "../../utils/promptUtils";
+import { Prompt } from "aiconfig";
+import useLoadModels from "../../hooks/useLoadModels";
+import { IconX } from "@tabler/icons-react";
+
+type Props = {
+  prompt: Prompt;
+  getModels: (search: string) => Promise<string[]>;
+  onSetModel: (model?: string) => void;
+  defaultConfigModelName?: string;
+};
+
+export default memo(function ModelSelector({
+  prompt,
+  getModels,
+  onSetModel,
+  defaultConfigModelName,
+}: Props) {
+  const [selectedModel, setSelectedModel] = useState<string | undefined>(
+    getPromptModelName(prompt, defaultConfigModelName)
+  );
+  const [showAll, setShowAll] = useState(true);
+  const [autocompleteSearch, setAutocompleteSearch] = useState(
+    getPromptModelName(prompt, defaultConfigModelName)
+  );
+
+  const models = useLoadModels(showAll ? "" : autocompleteSearch, getModels);
+
+  const onSelectModel = (model?: string) => {
+    setSelectedModel(model);
+    onSetModel(model);
+  };
+
+  return (
+    <Autocomplete
+      placeholder="Select model"
+      limit={100}
+      maxDropdownHeight={200}
+      rightSection={
+        selectedModel ? (
+          <Button
+            size="xs"
+            variant="subtle"
+            mr={10}
+            onClick={() => {
+              onSelectModel(undefined);
+              setShowAll(true);
+              setAutocompleteSearch("");
+            }}
+          >
+            <IconX size={10} />
+          </Button>
+        ) : null
+      }
+      filter={(searchValue: string, item: AutocompleteItem) => {
+        if (showAll) {
+          return true;
+        }
+
+        const modelName: string = item.value;
+        return modelName
+          .toLocaleLowerCase()
+          .includes(searchValue.toLocaleLowerCase().trim());
+      }}
+      data={models}
+      value={autocompleteSearch}
+      onChange={(value: string) => {
+        setAutocompleteSearch(value);
+        setShowAll(false);
+        onSelectModel(value);
+        models.some((model) => {
+          if (model === value) {
+            setShowAll(true);
```

Comment:
Ok just to clarify, this `onChange` only occurs when a value is selected from the autocomplete options? Does this also occur when user changes the search string?

Reply:
This is mainly copied from our model selector in Workbooks and is sort of a workaround to match our desired behaviour. By default, the component will only ever show results in the dropdown that match the text in the input box. So, if the aiconfig has model set to gpt-4 when you load it, the dropdown would only ever show models with gpt-4 substring.
This isn't what we want, though. The behaviour we want is:
- if the user is typing to 'search' for a model, then the dropdown should filter on that search substring
- otherwise, if the model is selected from the list / defaulted from the config then we want to show all the model options to change to

Reply:
Oh, I generally assumed that the user would just click the x button on the right to see everything, hmmmm. I don't think it's a blocker but feels a bit strange to me that auto-completions don't remain filtered after selected. We can do this as P2 someday